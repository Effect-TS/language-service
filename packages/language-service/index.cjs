// eslint-disable-next-line @typescript-eslint/consistent-type-imports, no-undef
module.exports = function init(modules) {
  function create(info) {
    // start with regular language service, then load custom once async module has loaded
    let languageService = info.languageService
    import("./build/cjs/index.js").then(v => languageService = v.default(modules).create(info))

    // create the proxy
    const proxy = Object.create(null)
    for (const k of Object.keys(info.languageService)) {
      proxy[k] = (...args) => languageService[k].apply(languageService, args)
    }

    return proxy
  }

  return { create }
}
