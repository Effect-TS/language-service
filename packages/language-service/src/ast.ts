import { Tag } from "@tsplus/stdlib/service/Tag"
import type ts from "typescript/lib/tsserverlibrary"

declare module "typescript/lib/tsserverlibrary" {
  const nullTransformationContext: ts.TransformationContext

  export namespace formatting {
    interface FormattingHost {
      getNewLine?(): string
    }

    export interface FormatContext {
      readonly options: ts.FormatCodeSettings
      readonly getRules: unknown
    }

    function getFormatContext(options: ts.FormatCodeSettings, host: FormattingHost): FormatContext
  }

  export type TextChangesContext = any
  export type ChangeNodeOptions = any

  export namespace textChanges {
    export class ChangeTracker {
      static with(context: ts.TextChangesContext, cb: (tracker: ChangeTracker) => void): ts.FileTextChanges[]
      replaceNode(sourceFile: ts.SourceFile, oldNode: ts.Node, newNode: ts.Node, options?: ts.ChangeNodeOptions): void
    }
    export function applyChanges(text: string, changes: readonly ts.TextChange[]): string
  }

  export function findPrecedingToken(
    position: number,
    sourceFile: ts.SourceFileLike,
    startNode: ts.Node,
    excludeJsdoc?: boolean
  ): ts.Node | undefined
  export function findPrecedingToken(
    position: number,
    sourceFile: ts.SourceFile,
    startNode?: ts.Node,
    excludeJsdoc?: boolean
  ): ts.Node | undefined

  export function isMemberName(node: ts.Node): node is ts.MemberName
  export function isKeyword(token: ts.SyntaxKind): token is ts.KeywordSyntaxKind
}

// eslint-disable-next-line @typescript-eslint/consistent-type-imports
export type TypeScriptApi = typeof import("typescript/lib/tsserverlibrary")
export const TypeScriptApi = Tag<TypeScriptApi>()
export const LanguageServiceApi = Tag<ts.LanguageService>()
export const ChangeTrackerApi = Tag<ts.textChanges.ChangeTracker>()

export class NoTypeScriptProgramError {
  readonly _tag = "NoTypeScriptProgramError"
}

export function getProgram() {
  return Effect.serviceWithEffect(
    LanguageServiceApi,
    languageService => Effect.sync(() => languageService.getProgram())
  )
    .filterOrFail((program): program is ts.Program => !!program, () => new NoTypeScriptProgramError())
}

export class NoSuchSourceFile {
  readonly _tag = "NoSuchSourceFile"
  constructor(
    readonly fileName: string
  ) {}
}

export function getSourceFile(fileName: string) {
  return getProgram().map(program => program.getSourceFile(fileName))
    .filterOrFail((sourceFile): sourceFile is ts.SourceFile => !!sourceFile, () => new NoSuchSourceFile(fileName))
}

export function hasModifier(node: ts.Declaration, kind: ts.ModifierFlags) {
  return Effect.serviceWith(TypeScriptApi, ts => !!(ts.getCombinedModifierFlags(node) & kind))
}

function getChildNodesContainingRange(
  parent: ts.Node,
  textRange: ts.TextRange
) {
  return Do($ => {
    const ts = $(Effect.service(TypeScriptApi))
    let result: Chunk<ts.Node> = Chunk.empty()

    ts.forEachChild(parent, node => {
      if (node.end < textRange.pos) {
        return
      } else if (node.pos > textRange.end) {
        return
      } else {
        result = result.append(node)
      }
    })

    return result
  })
}

/**
 * Gets the closest node that contains given TextRange
 */
export function getNodesContainingRange(
  sourceFile: ts.SourceFile,
  textRange: ts.TextRange
) {
  return Do(($) => {
    let result: Chunk<ts.Node> = Chunk.empty()

    $(
      Effect.iterate([sourceFile] as ts.Node[], (_) => _.length > 0)(toProcess => {
        const node = toProcess.shift()!
        return getChildNodesContainingRange(node, textRange).flatMap(_ =>
          Effect.sync(() => {
            result = result.concat(_)
            return toProcess.concat(..._.toArray)
          })
        )
      })
    )

    return Chunk.from(result.reverse)
  })
}

/**
 * Ensures value is a text range
 */
export function toTextRange(positionOrRange: number | ts.TextRange): ts.TextRange {
  return typeof positionOrRange === "number" ? { end: positionOrRange, pos: positionOrRange } : positionOrRange
}

export function getHumanReadableName(sourceFile: ts.SourceFile, node: ts.Node) {
  const text = node.getText(sourceFile)
  return text.length > 10 ? text.substring(0, 10) + "..." : text
}

export function collectAll<R, E>(rootNode: ts.Node, test: (node: ts.Node) => Effect<R, E, boolean>) {
  return Do(($) => {
    const ts = $(Effect.service(TypeScriptApi))
    const env = $(Effect.environment<R>())
    let result = Chunk.empty<ts.Node>()

    function visitor(node: ts.Node) {
      if (test(node).provideEnvironment(env).unsafeRunSync()) result = result.append(node)
      ts.visitEachChild(node, visitor, ts.nullTransformationContext)

      return node
    }

    ts.visitNode(rootNode, visitor)

    return result
  })
}

export function getRelevantTokens(
  position: number,
  sourceFile: ts.SourceFile
) {
  return Do(($) => {
    const ts = $(Effect.service(TypeScriptApi))
    const previousToken = ts.findPrecedingToken(position, sourceFile)
    if (
      previousToken && position <= previousToken.end &&
      (ts.isMemberName(previousToken) || ts.isKeyword(previousToken.kind))
    ) {
      const contextToken = ts.findPrecedingToken(previousToken.getFullStart(), sourceFile, /*startNode*/ undefined)! // TODO: GH#18217
      return { contextToken, previousToken }
    }
    return { contextToken: Maybe.fromNullable(previousToken), previousToken: Maybe.fromNullable(previousToken) }
  })
}
