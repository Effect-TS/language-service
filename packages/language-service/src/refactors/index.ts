import addPipe from "@effect/language-service/refactors/addPipe"
import asyncAwaitToGen from "@effect/language-service/refactors/asyncAwaitToGen"
import asyncAwaitToGenTryPromise from "@effect/language-service/refactors/asyncAwaitToGenTryPromise"
import functionToArrow from "@effect/language-service/refactors/functionToArrow"
import removeCurryArrow from "@effect/language-service/refactors/removeCurryArrow"
import removePipe from "@effect/language-service/refactors/removePipe"

export default { asyncAwaitToGen, asyncAwaitToGenTryPromise, removePipe, addPipe, removeCurryArrow, functionToArrow }
