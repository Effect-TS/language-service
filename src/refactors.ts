import { asyncAwaitToGen } from "./refactors/asyncAwaitToGen.js"
import { asyncAwaitToGenTryPromise } from "./refactors/asyncAwaitToGenTryPromise.js"
import { effectGenToFn } from "./refactors/effectGenToFn.js"
import { functionToArrow } from "./refactors/functionToArrow.js"
import { pipeableToDatafirst } from "./refactors/pipeableToDatafirst.js"
import { removeUnnecessaryEffectGen } from "./refactors/removeUnnecessaryEffectGen.js"
import { toggleLazyConst } from "./refactors/toggleLazyConst.js"
import { toggleReturnTypeAnnotation } from "./refactors/toggleReturnTypeAnnotation.js"
import { toggleTypeAnnotation } from "./refactors/toggleTypeAnnotation.js"
import { wrapWithEffectGen } from "./refactors/wrapWithEffectGen.js"
import { wrapWithPipe } from "./refactors/wrapWithPipe.js"

export const refactors = [
  asyncAwaitToGen,
  asyncAwaitToGenTryPromise,
  functionToArrow,
  pipeableToDatafirst,
  removeUnnecessaryEffectGen,
  toggleLazyConst,
  toggleReturnTypeAnnotation,
  toggleTypeAnnotation,
  wrapWithEffectGen,
  wrapWithPipe,
  effectGenToFn
]
