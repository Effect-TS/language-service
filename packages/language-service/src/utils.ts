import * as AST from "@effect/language-service/ast"
import type ts from "typescript/lib/tsserverlibrary"

export function isPipeCall(node: ts.Node) {
  return Do($ => {
    const ts = $(Effect.service(AST.TypeScriptApi))
    if (!ts.isCallExpression(node)) return false
    const expression = node.expression
    if (!ts.isIdentifier(expression)) return false
    if (expression.getText(node.getSourceFile()) !== "pipe") return false
    return true
  })
}

export function asPipeableCallExpression(node: ts.Node) {
  return Do($ => {
    const ts = $(Effect.service(AST.TypeScriptApi))
    // ensure the node is a call expression
    if (!ts.isCallExpression(node)) return Maybe.none
    // with just 1 arg
    if (node.arguments.length !== 1) return Maybe.none
    const arg = node.arguments[0]!
    // ideally T.map(n => n * 2) could be piped to pipe(n => n * 2, T.map)
    // but does not make any sense.
    if (ts.isArrowFunction(arg)) return Maybe.none
    // same goes for identifiers, string literal or numbers
    if (ts.isStringLiteral(arg) || ts.isNumericLiteral(arg) || ts.isIdentifier(arg)) return Maybe.none
    return Maybe.some([node.expression, arg] as const)
  })
}

export function asPipeArguments(initialNode: ts.Node) {
  return Do($ => {
    let result = Chunk.empty<ts.Expression>()
    $(
      Effect.iterate(Maybe.some(initialNode), node => node.isSome())(
        maybeNode =>
          Do($ => {
            if (maybeNode.isNone()) return Maybe.none

            const node = maybeNode.value
            const maybePipeable = $(asPipeableCallExpression(node))
            if (maybePipeable.isNone()) {
              result = result.append(node as ts.Expression)
              return Maybe.none
            }
            const [exp, arg] = maybePipeable.value
            result = result.append(exp)

            return Maybe.some(arg)
          })
      )
    )
    return result.reverse
  })
}

export function isPipeableCallExpression(node: ts.Node) {
  return asPipeableCallExpression(node).map(_ => _.isSome())
}

export function isCurryArrow(arrow: ts.Node) {
  return Do($ => {
    const ts = $(Effect.service(AST.TypeScriptApi))
    if (!ts.isArrowFunction(arrow)) return false
    if (arrow.parameters.length !== 1) return false
    const parameter = arrow.parameters[0]!
    const parameterName = parameter.name
    if (!ts.isIdentifier(parameterName)) return false
    const body = arrow.body
    if (!ts.isCallExpression(body)) return false
    const args = body.arguments
    if (args.length !== 1) return false
    const identifier = args[0]!
    if (!ts.isIdentifier(identifier)) return false
    return identifier.text === parameterName.text
  })
}

export function transformAsyncAwaitToEffectGen(
  node: ts.FunctionDeclaration | ts.ArrowFunction | ts.FunctionExpression,
  onAwait: (expression: ts.Expression) => ts.Expression
) {
  return Do($ => {
    const ts = $(Effect.service(AST.TypeScriptApi))

    function visitor(_: ts.Node): ts.Node {
      if (ts.isAwaitExpression(_)) {
        const expression = ts.visitEachChild(_.expression, visitor, ts.nullTransformationContext)

        return ts.factory.createYieldExpression(
          ts.factory.createToken(ts.SyntaxKind.AsteriskToken),
          ts.factory.createCallExpression(ts.factory.createIdentifier("$"), undefined, [onAwait(expression)])
        )
      }
      return ts.visitEachChild(_, visitor, ts.nullTransformationContext)
    }
    const generatorBody = visitor(node.body!)

    const generator = ts.factory.createFunctionExpression(
      [],
      ts.factory.createToken(ts.SyntaxKind.AsteriskToken),
      undefined,
      [],
      [ts.factory.createParameterDeclaration(undefined, undefined, "$")],
      undefined,
      generatorBody as any // NOTE(mattia): intended, to use same routine for both ConciseBody and Body
    )

    const effectGenCallExp = ts.factory.createCallExpression(
      ts.factory.createPropertyAccessExpression(
        ts.factory.createIdentifier("Effect"),
        "gen"
      ),
      undefined,
      [generator as any]
    )

    let currentFlags = ts.getCombinedModifierFlags(node)
    currentFlags &= ~ts.ModifierFlags.Async
    const newModifiers = ts.factory.createModifiersFromModifierFlags(currentFlags)

    if (ts.isArrowFunction(node)) {
      return ts.factory.createArrowFunction(
        newModifiers,
        node.typeParameters,
        node.parameters,
        undefined,
        node.equalsGreaterThanToken,
        effectGenCallExp
      )
    }

    const newBody = ts.factory.createBlock([
      ts.factory.createReturnStatement(effectGenCallExp)
    ])

    if (ts.isFunctionDeclaration(node)) {
      return ts.factory.createFunctionDeclaration(
        newModifiers,
        node.asteriskToken,
        node.name,
        node.typeParameters,
        node.parameters,
        undefined,
        newBody
      )
    }
    return ts.factory.createFunctionExpression(
      newModifiers,
      node.asteriskToken,
      node.name,
      node.typeParameters,
      node.parameters,
      undefined,
      newBody
    )
  })
}
