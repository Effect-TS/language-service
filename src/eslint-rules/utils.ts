/* eslint-disable @repo-tooling/dprint/dprint */
import { ESLintUtils } from "@typescript-eslint/utils"

const {getParserServices} = ESLintUtils

const createRule = ESLintUtils.RuleCreator(
  (name) => `https://example.com/rule/${name}`
)

export { getParserServices, createRule}
