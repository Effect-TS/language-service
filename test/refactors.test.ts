import * as T from "@effect/core/io/Effect"
import * as AST from "@effect/language-service/ast"
import type { RefactorDefinition } from "@effect/language-service/refactors/definition"
import refactors from "@effect/language-service/refactors/index"
import { createMockLanguageServiceHost } from "@effect/language-service/test/utils"
import { pipe } from "@tsplus/stdlib/data/Function"
import * as O from "@tsplus/stdlib/data/Maybe"
import * as fs from "fs"
import ts from "typescript/lib/tsserverlibrary"

/**
 * Loop through text changes, and update start and end positions while running
 */
function forEachTextChange(
  changes: ReadonlyArray<ts.TextChange>,
  cb: (change: ts.TextChange) => void
): void {
  changes = JSON.parse(JSON.stringify(changes))
  for (let i = 0; i < changes.length; i++) {
    const change = changes[i]!
    cb(change)
    const changeDelta = change.newText.length - change.span.length
    for (let j = i + 1; j < changes.length; j++) {
      if (changes[j]!.span.start >= change.span.start) {
        changes[j]!.span.start += changeDelta
      }
    }
  }
}

function applyEdits(edits: ReadonlyArray<ts.FileTextChanges>, fileName: string, sourceText: string): string {
  for (const fileTextChange of edits) {
    if (fileTextChange.fileName === fileName) {
      forEachTextChange(fileTextChange.textChanges, (edit) => {
        const content = sourceText
        const prefix = content.substring(0, edit.span.start)
        const middle = edit.newText
        const suffix = content.substring(edit.span.start + edit.span.length)
        sourceText = prefix + middle + suffix
      })
    }
  }
  return sourceText
}

export function testRefactorOnExample(refactor: RefactorDefinition, fileName: string) {
  const sourceWithMarker = fs.readFileSync(require.resolve(__dirname + "/../examples/refactors/" + fileName))
    .toString("utf8")
  const firstLine = (sourceWithMarker.split("\n")[0] || "").trim()
  for (const [textRangeString] of firstLine.matchAll(/([0-9]+:[0-9]+(-[0-9]+:[0-9]+)*)/gm)) {
    it(fileName + " at " + textRangeString, () => {
      // create the language service
      const sourceText = "// Result of running refactor " + refactor.name +
        " at position " + textRangeString + sourceWithMarker.substring(firstLine.length)
      const languageServiceHost = createMockLanguageServiceHost(fileName, sourceText)
      const languageService = ts.createLanguageService(languageServiceHost, undefined, ts.LanguageServiceMode.Semantic)
      const program = languageService.getProgram()
      if (!program) throw new Error("No typescript program!")
      const sourceFile = program.getSourceFile(fileName)
      if (!sourceFile) throw new Error("No source file " + fileName + " in VFS")

      // gets the position to test
      let startPos = 0
      let endPos = 0
      let i = 0
      for (const lineAndCol of textRangeString.split("-")) {
        const [line, character] = lineAndCol.split(":")
        const pos = ts.getPositionOfLineAndCharacter(sourceFile, +line! - 1, +character! - 1)
        if (i === 0) startPos = pos
        if (i === 1) endPos = pos
        i += 1
      }
      if (endPos < startPos) endPos = startPos
      const textRange = { pos: startPos, end: endPos }

      // ensure there are no errors in TS file
      const diagnostics = languageService.getCompilerOptionsDiagnostics()
        .concat(languageService.getSyntacticDiagnostics(fileName))
        .concat(languageService.getSemanticDiagnostics(fileName)).map((diagnostic) => {
          const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n")
          if (diagnostic.file) {
            const { character, line } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start!)
            return `  Error ${diagnostic.file.fileName} (${line + 1},${character + 1}): ${message}`
          } else {
            return `  Error: ${message}`
          }
        })
      expect(diagnostics).toEqual([])

      // check and assert the refactor is executable
      const canApply = pipe(
        refactor.apply(sourceFile, textRange),
        T.provideService(AST.TypeScriptApi, ts),
        T.provideService(AST.TypeScriptProgram, program),
        T.unsafeRunSync
      )

      if (O.isNone(canApply)) {
        expect(sourceText).toMatchSnapshot()
        return
      }

      // run the refactor and ensure it matches the snapshot
      const formatContext = ts.formatting.getFormatContext(
        ts.getDefaultFormatCodeSettings("\n"),
        { getNewLine: () => "\n" }
      )
      const edits = ts.textChanges.ChangeTracker.with(
        {
          formatContext,
          host: languageServiceHost,
          preferences: {}
        },
        (changeTracker) =>
          pipe(
            canApply.value.apply,
            T.provideService(AST.ChangeTrackerApi, changeTracker),
            T.provideService(AST.TypeScriptApi, ts),
            T.provideService(AST.TypeScriptProgram, program),
            T.unsafeRunSync
          )
      )

      expect(applyEdits(edits, fileName, sourceText)).toMatchSnapshot()
    })
  }
}

function testFiles(name: string, refactor: RefactorDefinition, fileNames: Array<string>) {
  for (const fileName of fileNames) {
    describe(fileName, () => {
      testRefactorOnExample(refactor, fileName)
    })
  }
}

Object.keys(refactors).map((refactorName) => testFiles(refactorName, refactors[refactorName]!, [refactorName + ".ts"]))
