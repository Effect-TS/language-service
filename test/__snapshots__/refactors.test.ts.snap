// Vitest Snapshot v1

exports[`addPipe.ts > addPipe.ts at 5:18 1`] = `
"// Result of running refactor effect/addPipe at position 5:18
import * as T from \\"@effect/io/Effect\\"
import { pipe } from \\"@fp-ts/data/Function\\"

const test = pipe(T.succeed(1), T.map((a: number) => a * 2))
"
`;

exports[`asyncAwaitToGen.ts > asyncAwaitToGen.ts at 4:28 1`] = `
"// Result of running refactor effect/asyncAwaitToGen at position 4:28
import * as T from \\"@effect/io/Effect\\"

export function refactorMe(arg: string) {
    return T.gen(function*($) {
        return yield* $(T.promise(Promise.resolve(1)))
    })
}
"
`;

exports[`asyncAwaitToGenTryPromise.ts > asyncAwaitToGenTryPromise.ts at 4:28 1`] = `
"// Result of running refactor effect/asyncAwaitToGenTryPromise at position 4:28
import * as Effect from \\"@effect/io/Effect\\"

export function refactorMe(arg: string) {
    return Effect.gen(function*($) {
        for (let i = 0; i < 10; i++) {
            yield* $(Effect.tryCatchPromise(() => Promise.resolve(i), error => ({ _tag: \\"Error1\\" as const, error })))
        }
        return yield* $(Effect.tryCatchPromise(() => Promise.resolve(arg), error => ({ _tag: \\"Error2\\" as const, error })))
    })
}
"
`;

exports[`functionToArrow.ts > functionToArrow.ts at 5:20 1`] = `
"// Result of running refactor effect/functionToArrow at position 5:20
/**
 * Docs
 */
export const myTest = (name: string) => name.length

/**
 * Docs
 */
export function myTest2(name: string) {
  if (name === \\"LOL\\") return 42
  return name.length
}

class Sample {
  /**
   * Docs
   */
  test(name: string) {
    return name.length
  }
}
"
`;

exports[`functionToArrow.ts > functionToArrow.ts at 12:20 1`] = `
"// Result of running refactor effect/functionToArrow at position 12:20
/**
 * Docs
 */
export function myTest(name: string) {
  return name.length
}

/**
 * Docs
 */
export const myTest2 = (name: string) => {
    if (name === \\"LOL\\")
        return 42
    return name.length
}

class Sample {
  /**
   * Docs
   */
  test(name: string) {
    return name.length
  }
}
"
`;

exports[`functionToArrow.ts > functionToArrow.ts at 21:5 1`] = `
"// Result of running refactor effect/functionToArrow at position 21:5
/**
 * Docs
 */
export function myTest(name: string) {
  return name.length
}

/**
 * Docs
 */
export function myTest2(name: string) {
  if (name === \\"LOL\\") return 42
  return name.length
}

class Sample {
  /**
   * Docs
   */
  /**
     * Docs
     */
    test = (name: string) => name.length
}
"
`;

exports[`noSyncWithConstant.ts > noSyncWithConstant.ts at 7:14 1`] = `
"// Result of running refactor effect/addPipe at position 7:14
import * as Effect from \\"@effect/io/Effect\\"
import { pipe } from \\"@fp-ts/data/Function\\"


const result =  pipe(
    Effect.succeed(\\"hello\\"),
    Effect.map((hello) => hello + \\", world!\\"),
    Effect.flatMap((msg) => Effect.log(msg))
  )
"
`;

exports[`removeCurryArrow.ts > removeCurryArrow.ts at 7:12 1`] = `
"// Result of running refactor effect/removeCurryArrow at position 7:12
import * as T from \\"@effect/io/Effect\\"
import { pipe } from \\"@fp-ts/data/Function\\"

const test = pipe(
  T.succeed(\\"Hello\\"),
  T.tap(T.log)
)
"
`;

exports[`removePipe.ts > removePipe.ts at 5:16 1`] = `
"// Result of running refactor effect/removePipe at position 5:16
import * as T from \\"@effect/io/Effect\\"
import { pipe } from \\"@fp-ts/data/Function\\"

const test = T.map(a => a * 1)(T.succeed(1))
"
`;

exports[`removePipe.ts > removePipe.ts at 7:7 1`] = `
"// Result of running refactor effect/removePipe at position 7:7
import * as T from \\"@effect/io/Effect\\"
import { pipe } from \\"@fp-ts/data/Function\\"

const test = pipe(
  1,
  T.succeed,
  T.map(a => a * 1)
)
"
`;

exports[`toggleReturnTypeAnnotation.ts > toggleReturnTypeAnnotation.ts at 4:32 1`] = `
"// Result of running refactor effect/toggleReturnTypeAnnotation at position 4:32
import * as T from \\"@effect/io/Effect\\"

export const test1 = (): T.Effect<never, never, number> => T.succeed(42)

export const test2 = () => (true ? T.succeed(42) : false)

function sillyGenerics<A>(value: A) {
  return T.fail(value)
}

function removeAnnotation():number{
  return 42
}

function removeAnnotationWithSpace(): number {
  return 42
}
"
`;

exports[`toggleReturnTypeAnnotation.ts > toggleReturnTypeAnnotation.ts at 6:32 1`] = `
"// Result of running refactor effect/toggleReturnTypeAnnotation at position 6:32
import * as T from \\"@effect/io/Effect\\"

export const test1 = () => T.succeed(42)

export const test2 = (): false | T.Effect<never, never, number> => (true ? T.succeed(42) : false)

function sillyGenerics<A>(value: A) {
  return T.fail(value)
}

function removeAnnotation():number{
  return 42
}

function removeAnnotationWithSpace(): number {
  return 42
}
"
`;

exports[`toggleReturnTypeAnnotation.ts > toggleReturnTypeAnnotation.ts at 8:32 1`] = `
"// Result of running refactor effect/toggleReturnTypeAnnotation at position 8:32
import * as T from \\"@effect/io/Effect\\"

export const test1 = () => T.succeed(42)

export const test2 = () => (true ? T.succeed(42) : false)

function sillyGenerics<A>(value: A): T.Effect<never, A, never> {
  return T.fail(value)
}

function removeAnnotation():number{
  return 42
}

function removeAnnotationWithSpace(): number {
  return 42
}
"
`;

exports[`toggleReturnTypeAnnotation.ts > toggleReturnTypeAnnotation.ts at 12:18 1`] = `
"// Result of running refactor effect/toggleReturnTypeAnnotation at position 12:18
import * as T from \\"@effect/io/Effect\\"

export const test1 = () => T.succeed(42)

export const test2 = () => (true ? T.succeed(42) : false)

function sillyGenerics<A>(value: A) {
  return T.fail(value)
}

function removeAnnotation(){
  return 42
}

function removeAnnotationWithSpace(): number {
  return 42
}
"
`;

exports[`toggleReturnTypeAnnotation.ts > toggleReturnTypeAnnotation.ts at 16:18 1`] = `
"// Result of running refactor effect/toggleReturnTypeAnnotation at position 16:18
import * as T from \\"@effect/io/Effect\\"

export const test1 = () => T.succeed(42)

export const test2 = () => (true ? T.succeed(42) : false)

function sillyGenerics<A>(value: A) {
  return T.fail(value)
}

function removeAnnotation():number{
  return 42
}

function removeAnnotationWithSpace() {
  return 42
}
"
`;

exports[`toggleTypeAnnotation.ts > toggleTypeAnnotation.ts at 4:16 1`] = `
"// Result of running refactor effect/toggleTypeAnnotation at position 4:16
import * as T from \\"@effect/io/Effect\\"

export const test1: <A>(value: A) => T.Effect<never, never, A> = T.succeed
export const test2 = T.fail(\\"LOL\\")

const predefined = 42
export const test3 = predefined

const callable = () => 42
export const test4 = callable
export const test5 = T.die

const removeAnnotation:number=42
const removeAnnotationWithSpace: number = 42
"
`;

exports[`toggleTypeAnnotation.ts > toggleTypeAnnotation.ts at 5:16 1`] = `
"// Result of running refactor effect/toggleTypeAnnotation at position 5:16
import * as T from \\"@effect/io/Effect\\"

export const test1 = T.succeed
export const test2: T.Effect<never, string, never> = T.fail(\\"LOL\\")

const predefined = 42
export const test3 = predefined

const callable = () => 42
export const test4 = callable
export const test5 = T.die

const removeAnnotation:number=42
const removeAnnotationWithSpace: number = 42
"
`;

exports[`toggleTypeAnnotation.ts > toggleTypeAnnotation.ts at 8:16 1`] = `
"// Result of running refactor effect/toggleTypeAnnotation at position 8:16
import * as T from \\"@effect/io/Effect\\"

export const test1 = T.succeed
export const test2 = T.fail(\\"LOL\\")

const predefined = 42
export const test3: 42 = predefined

const callable = () => 42
export const test4 = callable
export const test5 = T.die

const removeAnnotation:number=42
const removeAnnotationWithSpace: number = 42
"
`;

exports[`toggleTypeAnnotation.ts > toggleTypeAnnotation.ts at 11:16 1`] = `
"// Result of running refactor effect/toggleTypeAnnotation at position 11:16
import * as T from \\"@effect/io/Effect\\"

export const test1 = T.succeed
export const test2 = T.fail(\\"LOL\\")

const predefined = 42
export const test3 = predefined

const callable = () => 42
export const test4: () => number = callable
export const test5 = T.die

const removeAnnotation:number=42
const removeAnnotationWithSpace: number = 42
"
`;

exports[`toggleTypeAnnotation.ts > toggleTypeAnnotation.ts at 12:16 1`] = `
"// Result of running refactor effect/toggleTypeAnnotation at position 12:16
import * as T from \\"@effect/io/Effect\\"

export const test1 = T.succeed
export const test2 = T.fail(\\"LOL\\")

const predefined = 42
export const test3 = predefined

const callable = () => 42
export const test4 = callable
export const test5: (defect: unknown) => T.Effect<never, never, never> = T.die

const removeAnnotation:number=42
const removeAnnotationWithSpace: number = 42
"
`;

exports[`toggleTypeAnnotation.ts > toggleTypeAnnotation.ts at 14:16 1`] = `
"// Result of running refactor effect/toggleTypeAnnotation at position 14:16
import * as T from \\"@effect/io/Effect\\"

export const test1 = T.succeed
export const test2 = T.fail(\\"LOL\\")

const predefined = 42
export const test3 = predefined

const callable = () => 42
export const test4 = callable
export const test5 = T.die

const removeAnnotation=42
const removeAnnotationWithSpace: number = 42
"
`;

exports[`toggleTypeAnnotation.ts > toggleTypeAnnotation.ts at 15:16 1`] = `
"// Result of running refactor effect/toggleTypeAnnotation at position 15:16
import * as T from \\"@effect/io/Effect\\"

export const test1 = T.succeed
export const test2 = T.fail(\\"LOL\\")

const predefined = 42
export const test3 = predefined

const callable = () => 42
export const test4 = callable
export const test5 = T.die

const removeAnnotation:number=42
const removeAnnotationWithSpace = 42
"
`;

exports[`wrapWithPipe.ts > wrapWithPipe.ts at 2:13-2:26 1`] = `
"// Result of running refactor effect/wrapWithPipe at position 2:13-2:26
const txt = pipe(\\"Hello World\\")
const txt2 = \\"Hello\\"
"
`;

exports[`wrapWithPipe.ts > wrapWithPipe.ts at 3:14-3:21 1`] = `
"// Result of running refactor effect/wrapWithPipe at position 3:14-3:21
const txt = \\"Hello World\\"
const txt2 = pipe(\\"Hello\\")
"
`;
